.github/actions/ipsec-key-rotate/action.yaml

```diff
@@ -25,56 +19,39 @@ runs:
         job-name: conformance-ipsec-e2e-key-rotation
         extra-connectivity-test-flags: ${{ inputs.extra-connectivity-test-flags }}
         operation-cmd: |
-          KEYID=$(kubectl get secret -n kube-system cilium-ipsec-keys -o go-template --template={{.data.keys}} | base64 -d | grep -oP "^\d+")
-          if [[ $KEYID -ge 15 ]]; then KEYID=0; fi
-
-          if [[ "${{ inputs.key-algo }}" == "gcm(aes)" ]]; then
-            key="rfc4106(gcm(aes)) $(dd if=/dev/urandom count=20 bs=1 2> /dev/null | xxd -p -c 64) 128"
-          elif [[ "${{ inputs.key-algo }}" == "cbc(aes)" ]]; then
-            key="hmac(sha256) $(dd if=/dev/urandom count=32 bs=1 2> /dev/null| xxd -p -c 64) cbc(aes) $(dd if=/dev/urandom count=32 bs=1 2> /dev/null| xxd -p -c 64)"
-          else
-            echo "Invalid key type"; exit 1
-          fi
-          data="{\"stringData\":{\"keys\":\"$((($KEYID+1)))${{ inputs.key-type-two }} ${key}\"}}"
+          current_key=$(./cilium-cli encryption key-status)
+          echo "Rotating IPsec key (current $current_key)"
+          key_per_node=$([ "${{ inputs.key-type-two }}" == "+" ] && echo "true" || echo "false")
+          ./cilium-cli encryption rotate-key --auth-algo=${{ inputs.key-algo }} --key-per-node=$key_per_node
 
-          echo "Updating IPsec secret with $data"
-          kubectl patch secret -n kube-system cilium-ipsec-keys -p="$data" -v=1
-
-          # Compute number of expected keys during key rotation depending on
-          # whether we use the single-key system (1 key) or the per-tunnel
-          # keys system (4 keys).
-          exp_nb_keys=2
-          if [[ "${{ inputs.key-type-one }}" == "+" ]]; then
-            ((exp_nb_keys+=7))
-          fi
-          if [[ "${{ inputs.key-type-two }}" == "+" ]]; then
-            ((exp_nb_keys+=7))
-          fi
+          new_key=$(./cilium-cli encryption key-status)
+          echo "New $new_key"
 
           # Wait until key rotation starts
           # We expect the amount of keys in use to grow during rotation.
           while true; do
-            keys_in_use=$(kubectl -n kube-system exec daemonset/cilium -c cilium-agent -- cilium-dbg encrypt status | awk '/Keys in use/ {print $NF}')
-            if [[ $keys_in_use == $exp_nb_keys ]]; then
```

Comment:
This is quite nice from a user experience point of view, but in terms of CI, I believe we're losing coverage here.

In the current CI, if there is a bug that causes us not to use the proper key system (ex. use single key instead of per-tunnel keys), we will notice because the key rotation test will fail to proceed. With this new logic, it looks like we don't check the number of keys and will therefore proceed in case of such a bug. Or am I missing something?

Until we remove the single key system, I think it will be important to keep that check. It's the only place where we check this and we really don't want to regress here :sweat_smile:

Reply:
IPsec [key count logic](https://github.com/cilium/cilium-cli/blob/main/encrypt/status.go#L197-L212) implemented in Cilium CLI.

Reply:
Oh, perfect!

Reply:
My bad! I've improved the commit message.
Please let me know if it's not enough.
Thanks!